File: frontend/types/property.ts
import type { SeoMeta } from "@/types/page";
import type { Category } from "@/types/taxonomy";

export type BudgetBracket = "under2" | "2to5" | "5to10" | "10plus";
export type PropertyStatus = "draft" | "published";

export interface PropertyMediaItem {
  media_id: string;
  position: number;
}

export interface PropertyAmenity {
  name: string;
  image_media_id: string | null;
}

export interface PropertyListItem {
  id: string;
  name: string;
  slug: string;
  city: string | null;
  category_names: string[];
  price_display: string | null;
  status: PropertyStatus;
  sort_order: number;
  updated_at: string;
}

export interface Property {
  id: string;
  name: string;
  slug: string;
  city: string | null;
  location_text: string | null;
  price_display: string | null;
  price_value: string | null;
  budget_bracket: BudgetBracket | null;
  spec_a: string | null;
  spec_b: string | null;
  area_sqft: string | null;
  beds_options: string[] | null;
  description: string | null;
  amenities: PropertyAmenity[] | null;
  tag_text: string | null;
  status_text: string | null;
  is_signature: boolean;
  featured_image_media_id: string | null;
  brochure_media_id: string | null;
  seo: SeoMeta | null;
  status: PropertyStatus;
  sort_order: number;
  map_project_id: string | null;
  categories: Category[];
  gallery: PropertyMediaItem[];
  created_at: string;
  updated_at: string;

  // Extended Cinematic & Conversion Fields
  hero_media_type?: "video" | "image" | "carousel";
  desktop_hero_video_url?: string | null;
  mobile_hero_video_url?: string | null;
  desktop_hero_image_id?: string | null;
  mobile_hero_image_id?: string | null;
  poster_image_id?: string | null;
  hero_heading?: string | null;
  hero_subheading?: string | null;
  hero_overlay_strength?: number;
  hero_text_align?: "left" | "center" | "right";
  hero_theme?: "dark" | "light";
  video_experience?: any[];
  master_plan_media_id?: string | null;
  master_plan_title?: string | null;
  master_plan_description?: string | null;
  floor_plans?: any[];
  location_landmarks?: any[];
  highlights?: any[];
  offers?: any[];
  construction_updates?: any[];
  sections?: any[];
  rera_number?: string | null;
  approval_info?: string | null;
  disclaimer_text?: string | null;
  possession_date?: string | null;
}

File: backend/app/schemas/property.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.models.property import BudgetBracket, PropertyStatus
from app.schemas.seo import SeoMetaInput, SeoMetaRead
from app.schemas.taxonomy import CategoryRead
from app.schemas.validators import Slug


class PropertyMediaRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    media_id: str
    position: int


class PropertyAmenity(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    image_media_id: str | None = None


class PropertyListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    slug: str
    city: str | None
    category_names: list[str]
    price_display: str | None
    status: PropertyStatus
    sort_order: int
    updated_at: datetime


class PropertyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    slug: str
    city: str | None
    location_text: str | None
    price_display: str | None
    price_value: Decimal | None
    budget_bracket: BudgetBracket | None
    spec_a: str | None
    spec_b: str | None
    area_sqft: Decimal | None
    beds_options: list | None
    description: str | None
    amenities: list[PropertyAmenity] | None
    tag_text: str | None
    status_text: str | None
    is_signature: bool
    featured_image_media_id: str | None
    brochure_media_id: str | None
    map_project_id: str | None
    seo: SeoMetaRead | None
    status: PropertyStatus
    sort_order: int
    categories: list[CategoryRead]
    gallery: list[PropertyMediaRead]
    created_at: datetime
    updated_at: datetime


class PropertyCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    slug: Slug
    city: str | None = None
    location_text: str | None = None
    price_display: str | None = None
    price_value: Decimal | None = None
    budget_bracket: BudgetBracket | None = None
    spec_a: str | None = None
    spec_b: str | None = None
    area_sqft: Decimal | None = None
    beds_options: list[str] = []
    description: str | None = None
    amenities: list[PropertyAmenity] | None = None
    tag_text: str | None = None
    status_text: str | None = None
    is_signature: bool = False
    featured_image_media_id: str | None = None
    brochure_media_id: str | None = None
    map_project_id: str | None = None
    category_ids: list[str] = []


class PropertyUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = None
    slug: Slug | None = None
    city: str | None = None
    location_text: str | None = None
    price_display: str | None = None
    price_value: Decimal | None = None
    budget_bracket: BudgetBracket | None = None
    spec_a: str | None = None
    spec_b: str | None = None
    area_sqft: Decimal | None = None
    beds_options: list[str] | None = None
    description: str | None = None
    amenities: list[PropertyAmenity] | None = None
    tag_text: str | None = None
    status_text: str | None = None
    is_signature: bool | None = None
    featured_image_media_id: str | None = None
    brochure_media_id: str | None = None
    map_project_id: str | None = None
    category_ids: list[str] | None = None
    seo: SeoMetaInput | None = None


class PropertyGalleryRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    media_ids: list[str]


class PropertiesReorderRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order: list[str]
